Use SQL federation exec sql notExpression ClassCastException

## Bug Report

### Which version of ShardingSphere did you use?
master
### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
ShardingSphere-Proxy
### Expected behavior
![exeresult](https://github.com/apache/shardingsphere/assets/106047330/106b1d7c-2f12-469f-8200-416833499cc0)

### Actual behavior
```
[ERROR] 2023-12-15 17:08:16.593 [ShardingSphere-Command-0] o.a.s.p.f.c.CommandExecutorTask - Exception occur: 
java.lang.ClassCastException: class org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.NotExpression cannot be cast to class org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.simple.LiteralExpressionSegment (org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.NotExpression and org.apache.shardingsphere.sql.parser.sql.common.segment.dml.expr.simple.LiteralExpressionSegment are in unnamed module of loader 'app')
	at org.apache.shardingsphere.sql.parser.mysql.visitor.statement.MySQLStatementVisitor.createProjection(MySQLStatementVisitor.java:1775)
	at org.apache.shardingsphere.sql.parser.mysql.visitor.statement.MySQLStatementVisitor.visitProjection(MySQLStatementVisitor.java:1705)
	at org.apache.shardingsphere.sql.parser.mysql.visitor.statement.MySQLStatementVisitor.visitProjection(MySQLStatementVisitor.java:251)
	at org.apache.shardingsphere.sql.parser.autogen.MySQLStatementParser$ProjectionContext.accept(MySQLStatementParser.java:24357)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visit(AbstractParseTreeVisitor.java:18)
	at org.apache.shardingsphere.sql.parser.mysql.visitor.statement.MySQLStatementVisitor.visitProjections(MySQLStatementVisitor.java:1671)
	at org.apache.shardingsphere.sql.parser.mysql.visitor.statement.MySQLStatementVisitor.visitProjections(MySQLStatementVisitor.java:251)
	at org.apache.shardingsphere.sql.parser.autogen.MySQLStatementParser$ProjectionsContext.accept(MySQLStatementParser.java:23714)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visit(AbstractParseTreeVisitor.java:18)
	at org.apache.shardingsphere.sql.parser.mysql.visitor.statement.MySQLStatementVisitor.visitQuerySpecification(MySQLStatementVisitor.java:798)
	at org.apache.shardingsphere.sql.parser.mysql.visitor.statement.MySQLStatementVisitor.visitQuerySpecification(MySQLStatementVisitor.java:251)
	at org.apache.shardingsphere.sql.parser.autogen.MySQLStatementParser$QuerySpecificationContext.accept(MySQLStatementParser.java:21978)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visitChildren(AbstractParseTreeVisitor.java:46)
	at org.apache.shardingsphere.sql.parser.autogen.MySQLStatementBaseVisitor.visitQueryPrimary(MySQLStatementBaseVisitor.java:1399)
	at org.apache.shardingsphere.sql.parser.autogen.MySQLStatementParser$QueryPrimaryContext.accept(MySQLStatementParser.java:21895)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visit(AbstractParseTreeVisitor.java:18)
	at org.apache.shardingsphere.sql.parser.mysql.visitor.statement.MySQLStatementVisitor.visitQueryExpressionBody(MySQLStatementVisitor.java:761)
	at org.apache.shardingsphere.sql.parser.mysql.visitor.statement.MySQLStatementVisitor.visitQueryExpressionBody(MySQLStatementVisitor.java:251)
	at org.apache.shardingsphere.sql.parser.autogen.MySQLStatementParser$QueryExpressionBodyContext.accept(MySQLStatementParser.java:21608)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visit(AbstractParseTreeVisitor.java:18)
	at org.apache.shardingsphere.sql.parser.mysql.visitor.statement.MySQLStatementVisitor.visitQueryExpression(MySQLStatementVisitor.java:709)
	at org.apache.shardingsphere.sql.parser.mysql.visitor.statement.MySQLStatementVisitor.visitQueryExpression(MySQLStatementVisitor.java:251)
	at org.apache.shardingsphere.sql.parser.autogen.MySQLStatementParser$QueryExpressionContext.accept(MySQLStatementParser.java:21518)
	at org.antlr.v4.runtime.tree.AbstractParseTreeVisitor.visit(AbstractParseTreeVisitor.java:18)
	at org.apache.shardingsphere.sql.parser.mysql.visitor.statement.MySQLStatementVisitor.visitSelect(MySQLStatementVisitor.java:1625)
	at org.apache.shardingsphere.sql.parser.mysql.visitor.statement.MySQLStatementVisitor.visitSelect(MySQLStatementVisitor.java:251)
	at org.apache.shardingsphere.sql.parser.autogen.MySQLStatementParser$SelectContext.accept(MySQLStatementParser.java:21351)
	at org.apache.shardingsphere.sql.parser.api.SQLStatementVisitorEngine.visit(SQLStatementVisitorEngine.java:54)
	at org.apache.shardingsphere.infra.parser.sql.SQLStatementParserExecutor.parse(SQLStatementParserExecutor.java:47)
	at org.apache.shardingsphere.infra.parser.sql.SQLStatementParserEngine.parse(SQLStatementParserEngine.java:61)
	at org.apache.shardingsphere.infra.parser.ShardingSphereSQLParserEngine.parse(ShardingSphereSQLParserEngine.java:52)
	at org.apache.shardingsphere.proxy.frontend.mysql.command.query.text.query.MySQLComQueryPacketExecutor.parseSQL(MySQLComQueryPacketExecutor.java:80)
	at org.apache.shardingsphere.proxy.frontend.mysql.command.query.text.query.MySQLComQueryPacketExecutor.<init>(MySQLComQueryPacketExecutor.java:68)
	at org.apache.shardingsphere.proxy.frontend.mysql.command.MySQLCommandExecutorFactory.newInstance(MySQLCommandExecutorFactory.java:84)
	at org.apache.shardingsphere.proxy.frontend.mysql.command.MySQLCommandExecuteEngine.getCommandExecutor(MySQLCommandExecuteEngine.java:63)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.executeCommand(CommandExecutorTask.java:120)
	at org.apache.shardingsphere.proxy.frontend.command.CommandExecutorTask.run(CommandExecutorTask.java:78)
	at com.alibaba.ttl.TtlRunnable.run(TtlRunnable.java:60)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1136)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:635)
	at java.base/java.lang.Thread.run(Thread.java:833)
```
### Steps to reproduce the behavior, such as: SQL to execute, sharding rule configuration, when exception occur etc.
Environmental configuration Ref: #28895
```
select !0,NOT 0=1,!(0=0),1 AND 1,1 && 0,0 OR 1,1 || NULL, 1=1 or 1=1 and 1=0;
```

--- a/parser/sql/dialect/mysql/src/main/java/org/apache/shardingsphere/sql/parser/mysql/visitor/statement/MySQLStatementVisitor.java
+++ b/parser/sql/dialect/mysql/src/main/java/org/apache/shardingsphere/sql/parser/mysql/visitor/statement/MySQLStatementVisitor.java
@@ -1770,7 +1770,7 @@ public abstract class MySQLStatementVisitor extends MySQLStatementBaseVisitor<AS
             return projection;
         }
         if (projection instanceof CaseWhenExpression || projection instanceof VariableSegment || projection instanceof BetweenExpression || projection instanceof InExpression
-                || projection instanceof CollateExpression) {
+                || projection instanceof CollateExpression || projection instanceof NotExpression) {
             return createExpressionProjectionSegment(ctx, alias, projection);
         }
         LiteralExpressionSegment column = (LiteralExpressionSegment) projection;

--- a/test/it/parser/src/main/resources/case/dml/select.xml
+++ b/test/it/parser/src/main/resources/case/dml/select.xml
@@ -7525,4 +7525,149 @@
             </expr>
         </where>
     </select>
+    
+    <select sql-case-id="select_not_expression">
+        <projections start-index="7" stop-index="75">
+            <expression-projection text="!0" start-index="7" stop-index="8">
+                <expr>
+                    <not-expression start-index="7" stop-index="8">
+                        <expr>
+                            <literal-expression value="0" start-index="8" stop-index="8" />
+                        </expr>
+                    </not-expression>
+                </expr>
+            </expression-projection>
+            <expression-projection text="NOT 0=1" start-index="10" stop-index="16">
+                <expr>
+                    <binary-operation-expression start-index="10" stop-index="16">
+                        <left>
+                            <not-expression start-index="10" stop-index="14">
+                                <expr>
+                                    <literal-expression value="0" start-index="14" stop-index="14" />
+                                </expr>
+                            </not-expression>
+                        </left>
+                        <right>
+                            <literal-expression value="1" start-index="16" stop-index="16" />
+                        </right>
+                        <operator>=</operator>
+                    </binary-operation-expression>
+                </expr>
+            </expression-projection>
+            <expression-projection text="!(0=0)" start-index="18" stop-index="23">
+                <expr>
+                    <not-expression start-index="18" stop-index="23">
+                        <expr>
+                            <binary-operation-expression start-index="20" stop-index="22">
+                                <left>
+                                    <literal-expression value="0" start-index="20" stop-index="20" />
+                                </left>
+                                <right>
+                                    <literal-expression value="0" start-index="22" stop-index="22" />
+                                </right>
+                                <operator>=</operator>
+                            </binary-operation-expression>
+                        </expr>
+                    </not-expression>
+                </expr>
+            </expression-projection>
+            <expression-projection text="1 AND 1" start-index="25" stop-index="31">
+                <expr>
+                    <binary-operation-expression start-index="25" stop-index="31">
+                        <left>
+                            <literal-expression value="1" start-index="25" stop-index="25" />
+                        </left>
+                        <right>
+                            <literal-expression value="1" start-index="31" stop-index="31" />
+                        </right>
+                        <operator>AND</operator>
+                    </binary-operation-expression>
+                </expr>
+            </expression-projection>
+            <expression-projection text="1 &amp;&amp; 0" start-index="33" stop-index="38">
+                <expr>
+                    <binary-operation-expression start-index="33" stop-index="38">
+                        <left>
+                            <literal-expression value="1" start-index="33" stop-index="33" />
+                        </left>
+                        <right>
+                            <literal-expression value="0" start-index="38" stop-index="38" />
+                        </right>
+                        <operator>&amp;&amp;</operator>
+                    </binary-operation-expression>
+                </expr>
+            </expression-projection>
+            <expression-projection text="0 OR 1" start-index="40" stop-index="45">
+                <expr>
+                    <binary-operation-expression start-index="40" stop-index="45">
+                        <left>
+                            <literal-expression value="0" start-index="40" stop-index="40" />
+                        </left>
+                        <right>
+                            <literal-expression value="1" start-index="45" stop-index="45" />
+                        </right>
+                        <operator>OR</operator>
+                    </binary-operation-expression>
+                </expr>
+            </expression-projection>
+            <expression-projection text="1 || NULL" start-index="47" stop-index="55">
+                <expr>
+                    <binary-operation-expression start-index="47" stop-index="55">
+                        <left>
+                            <literal-expression value="1" start-index="47" stop-index="47" />
+                        </left>
+                        <right>
+                            <literal-expression value="null" start-index="52" stop-index="55" />
+                        </right>
+                        <operator>||</operator>
+                    </binary-operation-expression>
+                </expr>
+            </expression-projection>
+            <expression-projection text="1=1 or 1=1 and 1=0" start-index="58" stop-index="75">
+                <expr>
+                    <binary-operation-expression start-index="58" stop-index="75">
+                        <left>
+                            <binary-operation-expression start-index="58" stop-index="60">
+                                <left>
+                                    <literal-expression value="1" start-index="58" stop-index="58" />
+                                </left>
+                                <right>
+                                    <literal-expression value="1" start-index="60" stop-index="60" />
+                                </right>
+                                <operator>=</operator>
+                            </binary-operation-expression>
+                        </left>
+                        <right>
+                            <binary-operation-expression start-index="65" stop-index="75">
+                                <left>
+                                    <binary-operation-expression start-index="65" stop-index="67">
+                                        <left>
+                                            <literal-expression value="1" start-index="65" stop-index="65" />
+                                        </left>
+                                        <right>
+                                            <literal-expression value="1" start-index="67" stop-index="67" />
+                                        </right>
+                                        <operator>=</operator>
+                                    </binary-operation-expression>
+                                </left>
+                                <right>
+                                    <binary-operation-expression start-index="73" stop-index="75">
+                                        <left>
+                                            <literal-expression value="1" start-index="73" stop-index="73" />
+                                        </left>
+                                        <right>
+                                            <literal-expression value="0" start-index="75" stop-index="75" />
+                                        </right>
+                                        <operator>=</operator>
+                                    </binary-operation-expression>
+                                </right>
+                                <operator>and</operator>
+                            </binary-operation-expression>
+                        </right>
+                        <operator>or</operator>
+                    </binary-operation-expression>
+                </expr>
+            </expression-projection>
+        </projections>
+    </select>
 </sql-parser-test-cases>

--- a/test/it/parser/src/main/resources/sql/supported/dml/select.xml
+++ b/test/it/parser/src/main/resources/sql/supported/dml/select.xml
@@ -230,4 +230,5 @@
     <sql-case id="select_wm_concat_function4" value="SELECT NAME,WM_CONCAT(DECODE(SUBSTR((TO_CHAR(NAME)),0,1),'.','0'||TO_CHAR(NAME),TO_CHAR(NAME))) WM_NAME FROM TEST_TABLE WHERE NAME ='TEST' GROUP BY NAME" db-types="Oracle" />
     <sql-case id="select_with_user_updatable_columns" value="SELECT column_name, updatable FROM user_updatable_columns WHERE table_name = 'LOCATIONS_VIEW' ORDER BY column_name, updatable" db-types="Oracle" />
     <sql-case id="select_with_script_variables" value="SELECT x.$(ColumnName) FROM Person.Person x WHERE x.BusinessEntityID > 5" db-types="SQLServer" />
+    <sql-case id="select_not_expression" value="select !0,NOT 0=1,!(0=0),1 AND 1,1 &amp;&amp; 0,0 OR 1,1 || NULL, 1=1 or 1=1 and 1=0" db-types="MySQL" />
 </sql-cases>

